`requestAnimationFrame()` : Warn about long-lived applications and handle 0

Lectem opened the pull request:
### Description

Do not say that `requestAnimationFrame` returns non-zero values as this is not true, and warn about it.

### Motivation

For applications that may do a lot of calls to `requestAnimationFrame` (or that are long lived), this could cause reliability issues if users use 0 as an invalid handle.

### Additional details



For webkit:
- https://github.com/WebKit/WebKit/blob/6355462b3a9548b4d9df5a39afc5b4b923b57c95/Source/WebCore/dom/ScriptedAnimationController.h#L66 
- https://github.com/WebKit/WebKit/blob/6355462b3a9548b4d9df5a39afc5b4b923b57c95/Source/WebCore/dom/ScriptedAnimationController.cpp#L108


Exact same issue in Blink (since based on Webkit) 

- https://source.chromium.org/chromium/chromium/src/+/main:third_party/blink/renderer/core/dom/frame_request_callback_collection.cc;l=20;drc=27d34700b83f381c62e3a348de2e6dfdc08364b8

Since July 2024, Firefox detects overflow and will give an error: 
https://phabricator.services.mozilla.com/rMOZILLACENTRAL149722297f033d5c3ad126d0c72edcb1cb96d72e
https://searchfox.org/mozilla-central/source/dom/base/RequestCallbackManager.h#43

Lectem committed:
`requestAnimationFrame()` : Warn about long-lived applications and handle 0

For webkit:
https://github.com/WebKit/WebKit/blob/6355462b3a9548b4d9df5a39afc5b4b923b57c95/Source/WebCore/dom/ScriptedAnimationController.h#L66
https://github.com/WebKit/WebKit/blob/6355462b3a9548b4d9df5a39afc5b4b923b57c95/Source/WebCore/dom/ScriptedAnimationController.cpp#L108


Exact same issue in Blink (since based on Webkit) 

https://source.chromium.org/chromium/chromium/src/+/main:third_party/blink/renderer/core/dom/frame_request_callback_collection.cc;l=20;drc=27d34700b83f381c62e3a348de2e6dfdc08364b8

Lectem committed:
Update files/en-us/web/api/window/requestanimationframe/index.md

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

Josh-Cena commented on files/en-us/web/api/window/requestanimationframe/index.md:
```diff
@@ -50,10 +50,14 @@ requestAnimationFrame(callback)
 ### Return value
 
 A `long` integer value, the request ID, that uniquely identifies the entry
```
```suggestion
A `unsigned long` integer value, the request ID, that uniquely identifies the entry
```

If we want to talk about the wraparound behavior, it's better to say that it's at least non-negative.

Lectem replied:
Sadly this is not even the case! A simple signed integer is used. 
See https://github.com/WebKit/WebKit/blob/6355462b3a9548b4d9df5a39afc5b4b923b57c95/Source/WebCore/dom/ScriptedAnimationController.h#L66

It will wrap to negatives on virtually all modern CPU architectures(even through technically undefined behavior) then keep growing back until it hits 0 and starts the cycle again.

Lectem replied:
Actually, both Chromium and Webkit are wrong, but the spec indeed says `unsigned long`...
This was fixed only 6months ago for Firefox ! (and they mention not being able to make it 64bit due to the spec, probably because numbers in JS engines wouldn't be able to consume the value after a certain point.)
Note that only Firefox detects overflow! https://phabricator.services.mozilla.com/D216159

Josh-Cena commented on files/en-us/web/api/window/requestanimationframe/index.md:
```diff
@@ -50,10 +50,14 @@ requestAnimationFrame(callback)
 A `long` integer value, the request ID, that uniquely identifies the entry
-in the callback list. This is a non-zero value, but you may not make any other
-assumptions about its value. You can pass this value to
+in the callback list. You may not make any about its value. You can pass this value to
```
```suggestion
in the callback list. You should not make any assumptions about its value, other than it's a non-negative integer. You can pass this value to
```

Josh-Cena commented on files/en-us/web/api/window/requestanimationframe/index.md:
```diff
@@ -50,10 +50,14 @@ requestAnimationFrame(callback)
 {{domxref("window.cancelAnimationFrame()")}} to cancel the refresh callback request.
 
+> [!WARNING]
+> The returned value may seem to always be non-zero, however implementations typically increment an integer that may overflow and end up reaching 0.
+> While unlikely to cause issues for short-lived applications, you may want to use `null` for invalid handles instead of 0.
+> Note that it would take ~500 days to reach for a 60Hz based rendering with 100 calls to `requestAnimationFrame()` per frame.
```
```suggestion
> The request ID is typically implemented as a globally shared incrementing counter. Therefore, even when it starts counting at 1, it may overflow and end up reaching 0.
> While unlikely to cause issues for short-lived applications, you should avoid using `0` as a sentinel value for invalid handles, etc., and prefer unattainable values like `-1` or `null`.
> Note that it would take ~500 days to overflow the counter when rendering at 60Hz with 100 calls to `requestAnimationFrame()` per frame.
```

Lectem replied:
I'll reword a bit, however the counter is not shared but per window, and must be implemented as an incrementing counter.  https://html.spec.whatwg.org/multipage/imagebitmap-and-animations.html#animation-frame-callback-identifier

Josh-Cena reviewed (commented):
Just some minor suggestions for reword

Lectem committed:
Update RAF documentation with suggestions

Lectem committed:
RAF: mention firefox erroring out when max uint32_t is reached

Lectem committed:
Update files/en-us/web/api/window/requestanimationframe/index.md

Remove double space, though intended to go back to a new line, it is not allowed by the linter

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

Josh-Cena commented on files/en-us/web/api/window/requestanimationframe/index.md:
```diff
@@ -49,11 +49,17 @@ requestAnimationFrame(callback)
+> [!WARNING]
+> The request ID is typically implemented as a per-window incrementing counter. Therefore, even when it starts counting at 1, it may overflow and end up reaching 0.
+> While unlikely to cause issues for short-lived applications, you should avoid `0` as a sentinel value for invalid request identifier IDs and instead prefer unattainable values such as `null`.
+> As of January 2025, browsers such as WebKit and Chrome also don't implement it with an `unsigned long` but with a `signed long`, meaning you can not use negative values either. Firefox will return an error and fail instead of wrapping.
+> This also means that request IDs are not truely unique, and may be reused over time for browsers not implementing the check.
```
Let's be more handwavy here. The content isn't really for browser-specific behavior; if you want to, you would need to document them as implementation bugs in https://github.com/mdn/browser-compat-data.

```suggestion
> The spec doesn't specify the overflowing behavior, so browsers have divergent behaviors. When overflowing, the value would either wrap around to 0, to a negative value, or fail with an error.
> Unless overflow throws, request IDs are also not truly unique because there are only finitely many 32-bit integers for possibly infinitely many callbacks.
```

Josh-Cena reviewed (approved):
I would suggest that you create an issue and/or PR in https://github.com/mdn/browser-compat-data for browser-specific behaviors

Lectem committed:
Update files/en-us/web/api/window/requestanimationframe/index.md

Co-authored-by: Joshua Chen <sidachen2003@gmail.com>

Josh-Cena commented:
This looks great. Thanks!
